### ISSUE TITLE
Automated updates

### ISSUE THREAD

**Issue by ywarnier** [chamilo/chamilo-lms#5004]
Currently, Chamilo only offers the possibility to update by downloading a package and overwriting the existing code manually, then proceed through a few manual steps.

Ideally, we should have a secure process to have those updates downloaded and applied automatically (if the users so choose).

Here is a series of links that might help in that process:
- https://theupdateframework.io/
- https://github.com/php-tuf/php-tuf
- https://www.drupal.org/about/core/strategic-initiatives/automatic-updates

**Comment by ywarnier** [comment]
This issue is getting moved forward to 2.1 because we've freed some development time for this.
Ideally, to secude the update system, we should investigate a Cosign + Rekor (for transparency/audit) methodology, or alternatively use Minisign.
Given we'll only sign the whole Chamilo package for now, we shouldn't need an automation process at the beginning (this will only be needed once we start enabling plugins updates (after 3.0).

**Comment by christianbeeznest** [comment]
I opened an initial PR for this here: https://github.com/chamilo/chamilo-lms/pull/8531

## Summary

This PR adds the first safe layer for automated updates: **signed package verification and staging**.

It intentionally **does not apply updates yet**. No files are replaced, no migrations are executed, no maintenance mode is enabled, and no rollback logic is triggered.

The goal is to establish the secure verification/staging foundation before implementing the actual update application process.

## What is included

- **Update manifest reading**
  - channel
  - target version
  - release date
  - package URL
  - package SHA-256
  - signature type/URL
  - requirements such as PHP version

- **Package verification**
  - verifies that the package SHA-256 matches the manifest
  - rejects the package if the ZIP was modified or does not match the manifest

- **Optional Minisign signature verification**
  - verifies a package with Minisign using a trusted public key
  - works like an official seal: if the package was not signed with the expected private key, verification fails

- **Archive safety checks**
  - rejects unsafe ZIP entries such as path traversal or symlinks

- **Preflight checks**
  - update directory writable/creatable
  - disk space
  - PHP requirement
  - package type
  - version direction
  - Git working tree status

- **Staging extraction**
  - extracts the verified package into `var/update/staging/`
  - allows inspecting the package before any real update is applied

## Implemented flow

```text
check manifest
→ verify package
→ run preflight checks
→ prepare staging
```

## How to test locally

### 1. Create a minimal test package

```bash
cd /var/www/chamilo2

rm -rf /tmp/chamilo-stage-test /tmp/chamilo-stage-test.zip
mkdir -p /tmp/chamilo-stage-test/chamilo-test/{src,public,config}

cp composer.json /tmp/chamilo-stage-test/chamilo-test/composer.json
cp composer.lock /tmp/chamilo-stage-test/chamilo-test/composer.lock 2>/dev/null || true
cp yarn.lock /tmp/chamilo-stage-test/chamilo-test/yarn.lock 2>/dev/null || true

echo "test" > /tmp/chamilo-stage-test/chamilo-test/src/.keep
echo "test" > /tmp/chamilo-stage-test/chamilo-test/public/.keep

cd /tmp/chamilo-stage-test
zip -r /tmp/chamilo-stage-test.zip chamilo-test
sha256sum /tmp/chamilo-stage-test.zip
```

Copy the generated SHA-256.

### 2. Create a temporary manifest

```bash
cat > /tmp/update-manifest-stage-test.json <<'JSON'
{
  "channel": "stable",
  "version": "2.1.0",
  "released_at": "2026-01-20T00:00:00Z",
  "package": {
    "url": "https://download.chamilo.org/releases/chamilo-2.1.0.zip",
    "sha256": "REPLACE_WITH_REAL_SHA256"
  },
  "signature": {
    "type": "minisign",
    "url": "https://download.chamilo.org/releases/chamilo-2.1.0.zip.minisig"
  },
  "requirements": {
    "php": ">=8.2"
  }
}
JSON
```

Replace `REPLACE_WITH_REAL_SHA256` with the SHA-256 generated in step 1.

### 3. Test with CLI

```bash
php bin/console chamilo:update:check /tmp/update-manifest-stage-test.json

php bin/console chamilo:update:verify /tmp/update-manifest-stage-test.json \
  --package-path=/tmp/chamilo-stage-test.zip \
  --skip-signature

php bin/console chamilo:update:preflight /tmp/update-manifest-stage-test.json \
  --package-path=/tmp/chamilo-stage-test.zip

php bin/console chamilo:update:stage /tmp/update-manifest-stage-test.json \
  --package-path=/tmp/chamilo-stage-test.zip \
  --skip-signature
```

### 4. Test from the UI

Open:

```text
/admin/system-update
```

Use:

```text
Manifest source:
/tmp/update-manifest-stage-test.json

Package path:
/tmp/chamilo-stage-test.zip
```

Enable:

```text
Skip signature verification for local development tests
```

Then run:

- Check manifest
- Verify package
- Run preflight checks
- Prepare staging

Expected result:

- the manifest is loaded
- the package is verified
- preflight checks are completed
- the package is extracted into `var/update/staging/`
- no files are replaced
- no migrations are executed

## Next steps after this PR

Once this verification/staging layer is approved, the next work should be to:

- define the official manifest and signing-key location
- decide whether Minisign is enough for 2.1 or whether Cosign/Rekor should be added behind the same verifier layer
- add automatic package download from the manifest
- improve staging validation with a fuller dry-run report
- implement the real update application flow with maintenance mode, file replacement, migrations, and rollback

**Comment by christianbeeznest** [comment]
# GH-5004 Automated updates - current PR summary

This PR is a work in progress for #5004 and is not merged into `master` yet.

The current implementation adds the first end-to-end update flow for Chamilo 2.1+, starting from a signed/verified package and ending with controlled post-update commands. The goal is to let an administrator review and apply an update from the Chamilo administration UI while keeping each step explicit, auditable and reversible at the file level.

## Current scope

The updater currently supports this flow:

```text
Check manifest
→ Verify package
→ Run preflight checks
→ Prepare staging
→ Review apply plan
→ Apply staged files
→ Review post-apply actions
→ Review database migrations
→ Run selected post-apply actions
```

The implementation is intentionally step-based. Each stage can be reviewed before continuing, and the update does not automatically jump from download/verification to file replacement or database changes.

## Manifest and package verification

The update process starts from a manifest. The manifest describes the available release and the package metadata, including:

```json
{
  "channel": "stable",
  "version": "2.1.1",
  "released_at": "2026-06-05T00:00:00Z",
  "package": {
    "url": "https://download.chamilo.org/releases/chamilo-2.1.1.zip",
    "sha256": "..."
  },
  "signature": {
    "type": "minisign",
    "url": "https://download.chamilo.org/releases/chamilo-2.1.1.zip.minisig"
  },
  "requirements": {
    "php": ">=8.2"
  }
}
```

The updater verifies that the local or downloaded package matches the SHA-256 declared in the manifest. It also supports optional Minisign verification using a trusted public key.

For local development, the UI can use a local manifest and local ZIP package. For production, the same flow is expected to use the official Chamilo manifest and package URLs.


## Minisign server requirement

Minisign verification is executed on the Chamilo server during the package verification step. The browser does not verify the signature directly.

For production usage, the server must have the `minisign` command available to the PHP/CLI environment that runs the updater, and Chamilo must know the trusted public key used to verify official release packages. Only the public key is installed on the Chamilo server. The private signing key remains on the Chamilo release/signing infrastructure.

Example server setup on a Debian/Ubuntu-like system:

```bash
sudo apt-get update
sudo apt-get install minisign

minisign -v
```

Example verification command equivalent to what the updater performs internally:

```bash
minisign -Vm /path/to/chamilo-2.1.1.zip \
  -x /path/to/chamilo-2.1.1.zip.minisig \
  -P 'RW...trusted-public-key...'
```

In the updater flow, this happens after the manifest is read and the package path/signature path are known. If Minisign is not installed or the trusted public key is not configured, signature verification fails with a clear message. The local development UI can explicitly skip signature verification for test packages, but that option is not intended for production updates.

## Archive safety and staging

Before extracting a package, the updater performs archive safety checks. The package is rejected if unsafe ZIP entries are detected, such as path traversal entries or symbolic links.

After verification, the package is extracted into:

```text
var/update/staging/
```

This staging directory is separate from the active Chamilo installation. At this point, no application files have been replaced and no database changes have been executed.

## Preflight checks

The preflight step validates the environment before preparing an update. It currently checks items such as:

```text
- update directory availability
- disk space
- PHP version requirements
- local package readability
- project metadata
- Git working tree information when available
```

These checks are informational and safety-oriented. They help the administrator review the environment before continuing.

## Apply plan

The apply plan compares the staged package with the current Chamilo installation and builds a replacement plan.

It reports:

```text
- files that will be replaced
- new files that will be copied
- backup directory to be used
- update lock path
- target write permission checks
- symbolic link safety checks
- new staged Doctrine migration files
```

The plan is a review step only. It does not replace files.

## Applying staged files

After the administrator confirms the file update, the updater applies the staged files to the active Chamilo installation.

This step:

```text
- acquires an update lock
- backs up existing files planned for replacement
- copies staged files into the active installation
- writes an apply audit result
- records operation log entries for the live UI log
```

The file backup is stored under:

```text
var/update/backups/
```

The apply audit metadata is stored in the staging directory, allowing the administrator to inspect what was applied.

## Post-apply actions

After files are applied, the updater reviews the package and recommends post-apply actions when needed.

The currently supported controlled actions are:

```text
- Composer install
- Yarn install
- Yarn build
- Symfony cache clear
- Doctrine database migrations
```

The UI does not accept arbitrary shell commands. Only predefined, allowlisted actions can be executed.

Advanced actions, such as Composer, Yarn and database migrations, require explicit confirmation from the administrator.

## Runtime isolation for commands

Post-apply commands run from the current Chamilo project directory, but their runtime/cache directories are isolated inside the installation:

```text
var/update/runtime/
var/update/composer/
var/update/yarn/
var/update/npm/
```

This avoids relying on external home directories such as `/var/www` and keeps command runtime data inside the Chamilo update workspace.

## Database migration strategy

The updater is intended for Chamilo 2.1+ updates.

Because this updater starts existing in Chamilo 2.1, its database migration strategy is focused on future updates such as:

```text
2.1.0 → 2.1.1
2.1.1 → 2.1.2
2.1.x → later 2.1.x or compatible future versions
```

The updater does not run a global `doctrine:migrations:migrate` over the whole migration history. Instead, it detects new Doctrine migration files included in the staged package and executes those specific staged classes explicitly.

For Chamilo 2.1 updates, staged migrations are expected under:

```text
src/CoreBundle/Migrations/Schema/V210/
```

The execution model is:

```text
php bin/console doctrine:migrations:execute <staged-migration-class> --up --no-interaction
```

This avoids accidentally triggering historical migrations from older namespaces while still allowing an update package to include new schema changes.

Before running database migrations, the UI requires the administrator to confirm that a database backup exists and was verified. The updater currently does not create the database backup automatically.

## Example update package structure

A realistic update package is expected to contain source code and metadata, but not installation-specific runtime data.

Example:

```text
chamilo-2.1.1/
├── assets/
├── config/
├── public/
├── src/
├── templates/
├── translations/
├── composer.json
├── composer.lock
├── package.json
├── yarn.lock
└── webpack.config.js
```

It should not contain:

```text
.env
.env.local
vendor/
node_modules/
app/
var/cache/
var/log/
var/update/
var/upload/
var/plugins/
public/upload/
public/build/
.git/
```

Dependencies and generated assets are handled after the file update through the controlled post-apply actions.

## Example administrator flow

A typical administrator flow would be:

```text
1. Open the System update page from the administration panel.
2. Load the update manifest.
3. Verify the package hash and signature.
4. Run preflight checks.
5. Extract the package into staging.
6. Review the apply plan.
7. Confirm and apply staged files.
8. Review the recommended post-apply actions.
9. Review staged database migrations when present.
10. Confirm backup existence before running migrations.
11. Run selected post-apply actions.
12. Verify the resulting Chamilo installation.
```

This keeps the update process explicit and reviewable while moving the manual overwrite/update workflow into Chamilo's administration interface.

## Local development simulation

For development, the flow can be tested with a simulated package such as `2.1.1`, generated from a local Chamilo 2.1 working copy plus small additional changes, for example:

```text
- a new admin controller
- a new plugin directory
- a new V210 Doctrine migration
- updated composer metadata
- updated yarn metadata
- a visible file marker
```

This simulates how a future Chamilo release package would be applied to an installed Chamilo 2.1 instance.

## Summary

The current PR provides the foundation for automated updates in Chamilo 2.1:

```text
- manifest reading
- package hash verification
- optional Minisign verification
- safe ZIP extraction
- staging
- preflight checks
- apply plan review
- file backup and apply
- live operation logs
- post-apply action review
- controlled command execution
- staged V210 migration execution
```

The implementation remains review-first and does not silently modify code or database state without explicit administrator confirmation.

**Comment by ywarnier** [comment]
This is working. Great (and huge amount of) work! Hopefully this will start showing results soon (with the next minor version of Chamilo)
